### PR TITLE
fix: support JsonNullable in PATCH and add mapper regression tests

### LINKED ISSUE #94: [BUG] PATCH operations cannot clear fields (missing JsonNullable support in MapStruct)
### Description
Currently, PATCH operations cannot clear fields (e.g. setting `description` to `null`) because the `BaseEntityMapper.patchEntity` uses `NullValuePropertyMappingStrategy.IGNORE` and does not account for `JsonNullable`. This causes any `null` in the request body to be ignored instead of being applied to the entity.

### Current Implementation
In `src/main/java/com/budget/buddy/budget_buddy_api/base/crudl/base/BaseEntityMapper.java`:
```java
@BeanMapping(nullValuePropertyMappingStrategy = NullValuePropertyMappingStrategy.IGNORE)
// ...
void patchEntity(U patchRequest, @MappingTarget E existingEntity);
```

### Expected Behavior
MapStruct should be configured to handle `JsonNullable` (from the `jackson-databind-nullable` library) and check `JsonNullable.isPresent()` before applying updates. This would allow explicitly set `null` values to be propagated to the entity, while omitting fields would leave them unchanged.

### PR DESCRIPTION
**Pull request by glebremniov**
## Why

PATCH operations were unable to clear fields (setting to null) because MapStruct's IGNORE strategy was ignoring null results from JsonNullable conversion. This made it impossible to set a field like 'description' to null via the API.

Closes #94

## What changed

- **BaseEntityMapper.java**: Added `@Condition isPresent(JsonNullable<?>)` to handle property presence check specifically for JsonNullable types. This ensures regular fields in PUT/POST operations maintain standard mapping behavior (e.g., propagating nulls in full replacements).
- **BaseEntityMapper.java**: Configured `patchEntity` with `NullValuePropertyMappingStrategy.IGNORE` to skip omitted fields while allowing explicit nulls via JsonNullable.
- **TransactionMapperTest.java**: Added comprehensive regression tests for:
  - **PATCH**: Explicit nulls vs. undefined values using JsonNullable.
  - **PUT**: Full replacement behavior and verification that metadata fields (id, version, ownerId, createdAt, updatedAt) are strictly preserved.
- **CategoryMapperTest.java**: Added regression tests for full replacement and metadata preservation.
- **README.md**: Added a section explaining PATCH behavior and JSON Merge Patch semantics.
- **.gitignore**: Added `.junie/` to ignored files.

## Acceptance criteria

✅ MapStruct should handle JsonNullable from the jackson-databind-nullable library
✅ It should check JsonNullable.isPresent() before applying updates
✅ Explicitly set `null` values must be propagated to the entity in both PATCH and PUT
✅ Omitted fields in PATCH should leave the entity unchanged
✅ Identity and metadata fields must never be overridden during mapping

## How to verify

Run the mapper tests:
```bash
./gradlew test --tests com.budget.buddy.budget_buddy_api.transaction.TransactionMapperTest
./gradlew test --tests com.budget.buddy.budget_buddy_api.category.CategoryMapperTest
```

### CODE REVIEW

File: src/main/java/com/budget/buddy/budget_buddy_api/base/crudl/base/BaseEntityMapper.java [hunk 1]
```diff
@@ -97,4 +99,20 @@ public interface BaseEntityMapper<E extends BaseEntity<?>, R, C, U, L> {
...
+
+  /**
+   * Condition to check if a {@link JsonNullable} value is present.
+   * This is used by MapStruct for PATCH operations.
+   *
+   * @param value the property value from the source object
+   * @return true if the value is present (even if it's null), false otherwise
+   */
+  @Condition
+  default boolean isPresent(Object value) {
```
Comment (glebremniov) [inline src/main/java/com/budget/buddy/budget_buddy_api/base/crudl/base/BaseEntityMapper.java]: **Question**: `@Condition` on a method taking `Object` is a catch-all — MapStruct applies it as the presence check for **every** source property in all implementing mappers, not just `patchEntity`. For non-`JsonNullable` `null` values in non-patch methods (e.g. `toEntity`, `toResponse`), `isPresent(null)` returns `false` and MapStruct skips the property instead of mapping `null → null`.

In practice this is likely harmless (generated OpenAPI DTOs default nullable fields to `null`), but it's a subtle semantic change. Was this intentional? If so, a note in the Javadoc would help future readers understand why the fallback for non-`JsonNullable` types intentionally mirrors `NullValuePropertyMappingStrategy.IGNORE`.

File: src/main/java/com/budget/buddy/budget_buddy_api/base/crudl/base/BaseEntityMapper.java [hunk 2]
```diff
@@ -97,4 +99,20 @@ public interface BaseEntityMapper<E extends BaseEntity<?>, R, C, U, L> {
...
+    if (value instanceof JsonNullable<?> nullable) {
+      return nullable.isPresent();
+    }
+
+    return value != null;
```
Comment (glebremniov) [inline src/main/java/com/budget/buddy/budget_buddy_api/base/crudl/base/BaseEntityMapper.java]: **Nit**: The `return value != null` branch (the non-`JsonNullable` fallback) has no test coverage. The previous test `should_NotUpdate_If_Null` exercised this path via `update.setAmount(null)` (a plain `Long` field), but that assertion was removed in the refactor. Since this condition now controls ALL property mappings, consider adding a dedicated test that verifies a plain-null non-`JsonNullable` patch field is ignored.

File: src/test/java/com/budget/buddy/budget_buddy_api/transaction/TransactionMapperTest.java [hunk 3]
```diff
@@ -158,35 +158,48 @@ void should_UpdateOnlyProvidedFields() {
...
-          Currency.getInstance("EUR"), LocalDate.now(), "Old Description", UUID.randomUUID());
+          UUID.randomUUID(),
+          UUID.randomUUID(),
+          100L,
+          TransactionType.EXPENSE,
+          Currency.getInstance("EUR"),
+          LocalDate.now(),
+          "Old Description",
+          UUID.randomUUID()
+      );
```
Comment (glebremniov) [inline src/test/java/com/budget/buddy/budget_buddy_api/transaction/TransactionMapperTest.java:201]: **Praise**: The old `should_ClearDescription_When_ExplicitJsonNullableInPatch` test was silently wrong — it passed `JsonNullable.undefined()` while claiming to verify field clearing. This rename + corrected call to `JsonNullable.of(null)` fixes the test's semantics, not just the production code.

**Review by glebremniov — COMMENTED** [review]
Fixes the `JsonNullable` PATCH gap (#94) using MapStruct's `@Condition` mechanism. The implementation is minimal and well-targeted, and the test file also corrects a pre-existing bug where `should_ClearDescription_When_ExplicitJsonNullableInPatch` was actually passing `JsonNullable.undefined()` instead of `JsonNullable.of(null)` — good catch.

**Verdict: commented** — one question about scope of `@Condition`, one test-coverage suggestion. 3 inline comments.

File: src/main/java/com/budget/buddy/budget_buddy_api/base/crudl/base/BaseEntityMapper.java (hunk 1 above)
Comment (glebremniov) [reply src/main/java/com/budget/buddy/budget_buddy_api/base/crudl/base/BaseEntityMapper.java]: Addressed — isPresent now takes JsonNullable<?> directly, scoping the condition to only JsonNullable-typed properties. No longer affects toEntity/toResponse/etc. Resolving. ✓

File: src/main/java/com/budget/buddy/budget_buddy_api/base/crudl/base/BaseEntityMapper.java (hunk 2 above)
Comment (glebremniov) [reply src/main/java/com/budget/buddy/budget_buddy_api/base/crudl/base/BaseEntityMapper.java]: Addressed — the Object fallback branch is gone (isPresent now only handles JsonNullable<?>). The new ReplaceEntity tests in both TransactionMapperTest and CategoryMapperTest cover plain-null behavior for PUT operations. Resolving. ✓

File: src/test/java/com/budget/buddy/budget_buddy_api/transaction/TransactionMapperTest.java (hunk 3 above)
Comment (glebremniov) [reply src/test/java/com/budget/buddy/budget_buddy_api/transaction/TransactionMapperTest.java:201]: No action needed — this was praise. Resolving. ✓

File: src/test/java/com/budget/buddy/budget_buddy_api/transaction/TransactionMapperTest.java [hunk 4]
```diff
@@ -158,35 +159,49 @@ void should_UpdateOnlyProvidedFields() {
     }
 
     @Test
-    void should_NotUpdateIfNull() {
+    void
+    should_NotUpdate_When_Undefined() {
```
Comment (glebremniov) [inline src/test/java/com/budget/buddy/budget_buddy_api/transaction/TransactionMapperTest.java]: **Nit**: The method declaration is split across two lines (`void` on line 162, name on line 163). This is technically valid Java but looks like a formatter artifact. Should be `void should_NotUpdate_When_Undefined() {` on a single line to match the style of every other test method in this file.

**Review by glebremniov — COMMENTED** [review]
Re-review of d40fcfd.

**Addressed (2/2 actionable):**
- `@Condition` scope narrowed from `Object` to `JsonNullable<?>` — cleanly eliminates the catch-all concern and simplifies the implementation.
- Test coverage gap closed by new `ReplaceEntity` test classes in both `TransactionMapperTest` and `CategoryMapperTest`, which verify plain-null PUT behavior.

**New:** 1 formatting nit found (see inline).

**Verdict: commented** — 1 new inline comment, 3 prior threads resolved.
